components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1925,6 +1928,137 @@ void JsonRpcService::GetERC1155TokenBalance(
+  if (!network_url.is_valid()) {
+    std::move(callback).Run(
+        std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  base::Value topics(base::Value::Type::LIST);
+  if (!MakeAssetDiscoveryTopics(account_addresses, &topics)) {
+    std::move(callback).Run(
+        std::vector<mojom::BlockchainTokenPtr>(),
+        mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  // Create list of contract addresses to search by
+  // removing non erc20 tokens and assets that user has already
+  std::unordered_set<std::string> user_asset_contract_addresses;
+  for (auto& user_asset : user_assets) {
+    user_asset_contract_addresses.insert(user_asset->contract_address);
+  }
+  base::Value contract_addresses(base::Value::Type::LIST);
+  for (const auto& registry_token : token_registry) {
+    if (registry_token->is_erc20 && !registry_token->contract_address.empty() &&
+        user_asset_contract_addresses.find(registry_token->contract_address) ==
+            user_asset_contract_addresses.end()) {
+      contract_addresses.Append(registry_token->contract_address);
+    }
+  }
+
+  if (contract_addresses.GetList().size() == 0) {
+    std::move(callback).Run(std::vector<mojom::BlockchainTokenPtr>(),
+                            mojom::ProviderError::kSuccess, "");
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetTransferLogs,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  RequestInternal(
+      eth::eth_getLogs("earliest", "latest", &contract_addresses, &topics, ""),
+      true, network_url, std::move(internal_callback));
+  return;
+}
+
+void JsonRpcService::OnGetTransferLogs(
+    DiscoverAssetsCallback callback,
+    const int status,
+    const std::string& body,
+    const base::flat_map<std::string, std::string>& headers) {
+  std::vector<Log> logs;
+  if (!eth::ParseEthGetLogs(body, &logs)) {
+    std::move(callback).Run(std::vector<mojom::BlockchainTokenPtr>(),
+                            mojom::ProviderError::kParsingError,
+                            l10n_util::GetStringUTF8(IDS_WALLET_PARSING_ERROR));
+    return;
+  }
+
+  // Get unique list of token contract addresses with transfer events
+  std::unordered_set<std::string> contract_addresses;
+  for (const auto& log : logs) {
+    contract_addresses.insert(log.address);
+  }
+  std::vector<mojom::BlockchainTokenPtr> discovered_assets;
+
+  for (const auto& contract_address : contract_addresses) {
+    mojom::BlockchainTokenPtr token =
+        BlockchainRegistry::GetInstance()->GetTokenByAddress(
+            mojom::kMainnetChainId, mojom::CoinType::ETH, contract_address);
+    if (!token)
+      continue;
+    discovered_assets.push_back(std::move(token));
+  }
```

Comment:
I make this call to BlockchainRegistry::GetTokenByAddress in order to fetch the full BlockchainToken (not just the address).  Alternatively I could pass a map of contract_address -> BlockchainToken as an argument from OnGetAllTokensDiscoverAssets.  This map could replace the user_asset_contract_addresses unordered_set used in that function.

Reply:
I would prefer we pass along the map with tokens we got from GetAllTokens rather than getting token one by one again.